### ISSUE TITLE
ReadableStreamTee pulls too many chunks

### ISSUE THREAD

**Issue by MattiasBuelens** [whatwg/streams#996]
While reviewing #991, [I stumbled on a bug in `tee()`](https://github.com/whatwg/streams/issues/991#issuecomment-465814861).

For a single readable stream, we guarantee that we won't call `pull()` until the previous pull promise has resolved. But in the case of `tee()`, we use *the same pull algorithm* for both branches, so there's nothing stopping one branch from calling that function while the promise from a previous call by the other branch is still pending. Since both branches are constructed with a default high water mark of 1, they'll start out calling `pull()`, so we will start **2 read operations instead of just 1**.

## Reproduction code
```js
let remainingChunks = ['a', 'b', 'c'];
const rs = new ReadableStream({
  pull(c) {
    console.log('enqueue', remainingChunks[0]);
    c.enqueue(remainingChunks.shift());
    if (remainingChunks.length === 0) {
      console.log('close');
      c.close();
    }
  }
}, {highWaterMark: 0});
const [branch1, branch2] = rs.tee();
```
## Expected output
```
enqueue a
```
Both branches have HWM = 1, so `tee()` should read one chunk and enqueue it on both branches.

## Actual output
```
enqueue a
enqueue b
```
Both branches call `pull()` at the same time, so we read two chunks and enqueue them on both branches. This pushes them (unnecessarily) over their HWM = 1.

**Comment by MattiasBuelens** [comment]
I have a pretty good idea on how we can fix this. We'll have to borrow some of the logic from `ReadableStreamDefaultControllerCallPullIfNeeded` to check if we are already pulling, and return immediately if we are.

We can even choose whether we want to return the promise of the currently pending pull, or just `undefined`. If we return `undefined` without calling `enqueue()`, the branch won't call `pull()` again until we enqueue a new chunk, which also happens to be the moment when the currently pending pull promise resolves. 🙂

I haven't written a test case for WPT yet, but I can probably come up with something.

**Comment by domenic** [comment]
Be careful with words like "simultaneously" or "at the same time" in JS, since that's not really possible in a single-threaded language. I guess you mean "while the previous call's promise hasn't yet settled".

But yeah, otherwise, very nice find. Your OP should be pretty easy to turn into a web platform test, and illustrates the problem nicely.

**Comment by MattiasBuelens** [comment]
> Be careful with words like "simultaneously" or "at the same time" in JS, since that's not really possible in a single-threaded language. I guess you mean "while the previous call's promise hasn't yet settled".

Right, that's indeed what I meant. I'll update it for clarity.

**Comment by ricea** [comment]
I've uploaded tests at https://github.com/web-platform-tests/wpt/pull/15521. I have a fix in progress, but I'm not sure I have enough test coverage to be completely confident in it yet.

**Comment by MattiasBuelens** [comment]
I've also implemented a fix, see #997. Maybe we can compare notes? 😄 

I still want to come up with a few more test cases for this though...
